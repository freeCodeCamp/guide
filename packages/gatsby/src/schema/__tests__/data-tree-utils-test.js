const {
  extractFieldExamples,
  buildFieldEnumValues,
  INVALID_VALUE,
} = require(`../data-tree-utils`)

describe(`Gatsby data tree utils`, () => {
  const nodes = [
    {
      name: `The Mad Max`,
      hair: 1,
      date: `2006-07-22T22:39:53.000Z`,
      "key-with..unsupported-values": true,
      emptyArray: [],
      anArray: [1, 2, 3, 4],
      frontmatter: {
        date: `2006-07-22T22:39:53.000Z`,
        title: `The world of dash and adventure`,
        blue: 100,
      },
    },
    {
      name: `The Mad Wax`,
      hair: 2,
      date: `2006-07-22T22:39:53.000Z`,
      emptyArray: [undefined, null],
      anArray: [1, 2, 5, 4],
      iAmNull: null,
      frontmatter: {
        date: `2006-07-22T22:39:53.000Z`,
        title: `The world of slash and adventure`,
        blue: 10010,
        circle: `happy`,
        draft: false,
      },
    },
    {
      name: `The Mad Wax`,
      hair: 3,
      date: `2006-07-22T22:39:53.000Z`,
      anArray: [],
      iAmNull: null,
      frontmatter: {
        date: `2006-07-22T22:39:53.000Z`,
        title: `The world of slash and adventure`,
        blue: 10010,
        circle: `happy`,
        draft: false,
      },
    },
    {
      name: `The Mad Wax`,
      hair: 4,
      date: `2006-07-22T22:39:53.000Z`,
      anArray: [4, 6, 2],
      iAmNull: null,
      frontmatter: {
        date: `2006-07-22T22:39:53.000Z`,
        title: `The world of slash and adventure`,
        blue: 10010,
        circle: `happy`,
        draft: false,
      },
    },
  ]

  it(`builds field examples from an array of nodes`, () => {
    expect(extractFieldExamples(nodes)).toMatchSnapshot()
  })

  it(`null fields should have a null value`, () => {
    expect(extractFieldExamples(nodes).iAmNull).toBeNull()
  })

  it(`turns empty or sparse arrays to null`, () => {
    expect(extractFieldExamples(nodes).emptyArray).toBeNull()
    expect(extractFieldExamples(nodes).hair).toBeDefined()
  })

  it(`build enum values for fields from array on nodes`, () => {
    expect(buildFieldEnumValues(nodes)).toMatchSnapshot()
  })

  it(`turns polymorphic fields null`, () => {
    let example = extractFieldExamples([
      { foo: null },
      { foo: [1] },
      { foo: { field: 1 } },
    ])
    expect(example.foo).toBe(INVALID_VALUE)
  })

  it(`handles polymorphic arrays`, () => {
    let example = extractFieldExamples([
      { foo: [[`foo`, `bar`]] },
      { foo: [{ field: 1 }] },
    ])
    expect(example.foo).toBe(INVALID_VALUE)
  })

  it(`doesn't confuse empty fields for polymorhpic ones`, () => {
    let example = extractFieldExamples([
      { foo: { bar: 1 } },
      { foo: null },
      { foo: { field: 1 } },
    ])
    expect(example.foo).toEqual({ field: 1, bar: 1 })

    example = extractFieldExamples([
      { foo: [{ bar: 1 }] },
      { foo: null },
      { foo: [{ field: 1 }, { baz: 1 }] },
    ])
    expect(example.foo).toEqual([{ field: 1, bar: 1, baz: 1 }])
  })
})
