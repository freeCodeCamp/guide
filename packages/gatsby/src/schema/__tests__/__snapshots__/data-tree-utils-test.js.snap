// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gatsby data tree utils build enum values for fields from array on nodes 1`] = `
Object {
  "anArray": Object {
    "field": "anArray",
  },
  "date": Object {
    "field": "date",
  },
  "frontmatter___blue": Object {
    "field": "frontmatter.blue",
  },
  "frontmatter___circle": Object {
    "field": "frontmatter.circle",
  },
  "frontmatter___date": Object {
    "field": "frontmatter.date",
  },
  "frontmatter___draft": Object {
    "field": "frontmatter.draft",
  },
  "frontmatter___title": Object {
    "field": "frontmatter.title",
  },
  "hair": Object {
    "field": "hair",
  },
  "key___with______unsupported___values": Object {
    "field": "key-with..unsupported-values",
  },
  "name": Object {
    "field": "name",
  },
}
`;

exports[`Gatsby data tree utils builds field examples from an array of nodes 1`] = `
Object {
  "anArray": Array [
    1,
  ],
  "date": "2006-07-22T22:39:53.000Z",
  "emptyArray": null,
  "frontmatter": Object {
    "blue": 10010,
    "circle": "happy",
    "date": "2006-07-22T22:39:53.000Z",
    "draft": false,
    "title": "The world of slash and adventure",
  },
  "hair": 4,
  "iAmNull": null,
  "key-with..unsupported-values": true,
  "name": "The Mad Wax",
}
`;
